vlasovsolver/cpu_trans_pencils.cpp

```diff
@@ -0,0 +1,1170 @@
+// use DCCRG version Nov 8th 2018 01482cfba8
+#include "../grid.h"
+using namespace std;
+using namespace spatial_cell;
+
+#include "cpu_trans_pencils.hpp"
+
+std::array<setOfPencils,3> DimensionPencils;
+std::array<std::unordered_set<CellID>,3> DimensionTargetCells;
+
+//Is cell translated? It is not translated if DO_NO_COMPUTE or if it is sysboundary cell and not in first sysboundarylayer
+bool do_translate_cell(SpatialCell* SC){
+   if(SC->sysBoundaryFlag == sysboundarytype::DO_NOT_COMPUTE ||
+      (SC->sysBoundaryLayer != 1 && SC->sysBoundaryFlag != sysboundarytype::NOT_SYSBOUNDARY))
+      return false;
+   else
+      return true;
+}
+
+/* Get the one-dimensional neighborhood index for a given direction and neighborhood size.
+ *
+ * @param dimension spatial dimension of neighborhood
+ * @param stencil neighborhood size in cells
+ * @return neighborhood index that can be passed to DCCRG functions
+ */
+int getNeighborhood(const uint dimension, const uint stencil) {
+
+   int neighborhood = 0;
+
+   if (stencil == 1) {
+      switch (dimension) {
+      case 0:
+         neighborhood = VLASOV_SOLVER_TARGET_X_NEIGHBORHOOD_ID;
+         break;
+      case 1:
+         neighborhood = VLASOV_SOLVER_TARGET_Y_NEIGHBORHOOD_ID;
+         break;
+      case 2:
+         neighborhood = VLASOV_SOLVER_TARGET_Z_NEIGHBORHOOD_ID;
+         break;
+      }
+   }
+
+   if (stencil > 1) {
+      switch (dimension) {
+      case 0:
+         neighborhood = VLASOV_SOLVER_X_NEIGHBORHOOD_ID;
+         break;
+      case 1:
+         neighborhood = VLASOV_SOLVER_Y_NEIGHBORHOOD_ID;
+         break;
+      case 2:
+         neighborhood = VLASOV_SOLVER_Z_NEIGHBORHOOD_ID;
+         break;
+      }
+   }
+
+   return neighborhood;
+
+}
+
+void flagSpatialCellsForAmrCommunication(const dccrg::Dccrg<SpatialCell,dccrg::Cartesian_Geometry>& mpiGrid,
+                                         const vector<CellID>& localPropagatedCells) {
+
+   // Only flag/unflag cells if AMR is active
+   if (mpiGrid.get_maximum_refinement_level()==0) return;
+
+   // return if there's no cells to flag
+   if(localPropagatedCells.size() == 0) {
+      std::cerr<<"No cells!"<<std::endl;
```

Comment:
Should we take this opportunity and make this more explicit? Every now and then I run into this message during some tests and I'm worried and dig up this part of the code to check.

Reply:
ok, can make it more wordy.